### ISSUE TITLE
Quantum Dropout implementation

### ISSUE THREAD

**Issue by fran-scala** [PennyLaneAI/pennylane#4929]
### Feature details

Hi Pennylane developers! I'm opening this feature request to propose the implementation of the novel ***Quantum Dropout technique*** in Pennylane. This feature request is motivated by the findings and insights presented in our recent paper titled ["A General Approach to Dropout in Quantum Neural Networks"](https://arxiv.org/abs/2310.04120) (arXiv:2310.04120) that will soon be published in [Advanced Quantum Technologies](https://onlinelibrary.wiley.com/journal/25119044). In our research, we explored the application of dropout techniques in Quantum Neural Networks(QNNs) to mitigate potential overfitting issues arising from the overparametrization of quantum circuits.

Given the rising interest in overparametrized QNNs from the QML community, it will be vital to have regularization techniques. The proposed quantum dropout technique aims to provide Pennylane users with a mechanism to reduce overfitting, enhance generalization, and guide model training in QML applications. 

# Quantum Dropout Scheme

The dropout technique for QNNs follows a few simple steps:
1. While designing your ansatz, choose which gates will undergo dropout during the training procedure.
2. Fix a probability of dropout, i.e. the probability of removing one of the gates chosen in step 1. This can be defined locally, layerwise or globally.
Then at each training iteration:
3. Sample which gates have to be removed according to the probability defined in step 2. 
4. Build the corresponding QNN.
5. Perform the parameters update according to a certain optimizer.

### Implementation

# Rough implementation for the paper

For the extensive numerical analysis conducted in our paper, we employed Pennylane combined with JIT compilation of JAX.
Dropout essentially consists of ***randomly substituting a quantum gate with the identity***. For parametrized gates, we reduced this to conditionally set the parameter to 0, while for CNOT gate we had to define a custom parametrized gate.

# Suggested implementation

We suggest implementing this technique by ***adding a new parameter*** to all the gate functions which represents the ***dropout probability*** (with default value 0). Then according to this probability, the `compute_matrix` method returns either the gate matrix or the identity matrix (if dropout is applied).

It will be necessary that this works also in combination with JAX, which usually does not really like conditional statements.

### How important would you say this feature is?

1: Not important. Would be nice to have.

### Additional information

# Conclusion

You can find more details about the technique in our preprint ["A General Approach to Dropout in Quantum Neural Networks"](https://arxiv.org/abs/2310.04120) (arXiv:2310.04120). Do not hesitate to contact me at francesco.scala@ateneopv.it if you need some insights or discussion.

We believe that incorporating this technique into Pennylane would not only enhance the library's capabilities but also contribute to the advancement of QML research empowering researchers and practitioners. We also hope that it will become a ***standard practice for training deep QNNs***, just like its classical counterpart.

**Comment by isaacdevlugt** [comment]
Hey @fran-scala, thanks for the suggestion! We'll get back to you next week when we have a chance to digest the suggestions completely, but would this be something that you would like to contribute to PennyLane by opening a PR? Let us know!

**Comment by fran-scala** [comment]
> Hey @fran-scala, thanks for the suggestion! We'll get back to you next week when we have a chance to digest the suggestions completely, but would this be something that you would like to contribute to PennyLane by opening a PR? Let us know!

Yes, I would be delighted to contribute! However, I think I will still need some support to design the implementation correctly.

**Comment by josh146** [comment]
Hi @fran-scala! Before designing an implementation, another option could be to write a demonstration on your paper for https://pennylane.ai/qml/demonstrations/. That is, a Jupyter notebook that showcases the quantum dropout implementation, using code to demonstrate while explaining the concepts.

The nice thing about this is it allows us to understand the implementation better (helping with future design questions), while giving a spot for us to market your research/share it with the community :)

**Comment by fran-scala** [comment]
> Hi @fran-scala! Before designing an implementation, another option could be to write a demonstration on your paper for https://pennylane.ai/qml/demonstrations/. That is, a Jupyter notebook that showcases the quantum dropout implementation, using code to demonstrate while explaining the concepts.
> 
> The nice thing about this is it allows us to understand the implementation better (helping with future design questions), while giving a spot for us to market your research/share it with the community :)

Hi @josh146 ! Thanks for your proposal. The idea of making a demonstration notebook to showcase our technique sounds great and actually it should be pretty easy to do from our code. Do I have to notify you here when I submit the demo?

**Comment by josh146** [comment]
> Do I have to notify you here when I submit the demo?

@fran-scala nope!

To submit the demo, a pull request simply needs to be made against the https://github.com/pennylaneai/qml repository. Check out the [README](https://github.com/pennylaneai/qml#adding-demos) for more details on how to add a demo 🙂 

A couple of notes:

- Our website doesn't use notebooks as the demo format; instead, demos are Python scripts with [RestructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) comments. However, if you would like to start off with a notebook, you can convert your notebook to a Python script using [this converter script](https://github.com/PennyLaneAI/qml/tree/master/notebook_converter)

- Feel free to open a WIP (work-in-progress) PR against that repo, even if your demo is still a draft! This way we can help with any technical issues/answer any questions you might have. In addition, within each PR we have a GitHub action that will build a preview of your demo, making it convenient to check if everything is rendering correctly.

Let me know if you have any questions!

**Comment by fran-scala** [comment]
> > Do I have to notify you here when I submit the demo?
> 
> @fran-scala nope!
> 
> To submit the demo, a pull request simply needs to be made against the https://github.com/pennylaneai/qml repository. Check out the [README](https://github.com/pennylaneai/qml#adding-demos) for more details on how to add a demo 🙂
> 
> A couple of notes:
> 
> * Our website doesn't use notebooks as the demo format; instead, demos are Python scripts with [RestructuredText](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) comments. However, if you would like to start off with a notebook, you can convert your notebook to a Python script using [this converter script](https://github.com/PennyLaneAI/qml/tree/master/notebook_converter)
> * Feel free to open a WIP (work-in-progress) PR against that repo, even if your demo is still a draft! This way we can help with any technical issues/answer any questions you might have. In addition, within each PR we have a GitHub action that will build a preview of your demo, making it convenient to check if everything is rendering correctly.
> 
> Let me know if you have any questions!

Hi! I made a [pull request](https://github.com/PennyLaneAI/qml/pull/1003) in the qml module as you suggested. Let me know if it can work.

**Comment by isaacdevlugt** [comment]
Hey @fran-scala! It looks _awesome_! We might be a bit delayed getting to it because of the holidays, but I assure you that we will take a look at it as soon as we can. Let's move the conversation over to your PR that you made 😄
